storm-core/src/jvm/backtype/storm/scheduler/resource/strategies/eviction/DefaultEvictionStrategy.java

```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Override
+    public boolean makeSpaceForTopo(TopologyDetails td) {
+        LOG.debug("attempting to make space for topo {} from user {}", td.getName(), td.getTopologySubmitter());
+        User submitter = this.userMap.get(td.getTopologySubmitter());
+        if (submitter.getCPUResourceGuaranteed() == null || submitter.getMemoryResourceGuaranteed() == null) {
+            return false;
+        }
+        double cpuNeeded = td.getTotalRequestedCpu() / submitter.getCPUResourceGuaranteed();
+        double memoryNeeded = (td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap()) / submitter.getMemoryResourceGuaranteed();
+
+        User evictUser = this.findUserWithHighestAverageResourceUtilAboveGuarantee();
+        //user has enough resource under his or her resource guarantee to schedule topology
+        if ((1.0 - submitter.getCPUResourcePoolUtilization()) >= cpuNeeded && (1.0 - submitter.getMemoryResourcePoolUtilization()) >= memoryNeeded) {
+            if (evictUser != null) {
+                TopologyDetails topologyEvict = evictUser.getRunningTopologyWithLowestPriority();
+                evictTopology(topologyEvict);
+                return true;
+            }
+        } else {
+            if (evictUser != null) {
+                if ((evictUser.getResourcePoolAverageUtilization() - 1.0) > (cpuNeeded + (submitter.getResourcePoolAverageUtilization() - 1.0))) {
+                    TopologyDetails topologyEvict = evictUser.getRunningTopologyWithLowestPriority();
+                    evictTopology(topologyEvict);
+                    return true;
+                }
+            }
+        }
+        //See if there is a lower priority topology that can be evicted from the current user
+        for (TopologyDetails topo : submitter.getTopologiesRunning()) {
+            //check to if there is a topology with a lower priority we can evict
+            if (topo.getTopologyPriority() > td.getTopologyPriority()) {
+                evictTopology(topo);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void evictTopology(TopologyDetails topologyEvict) {
+        Collection<WorkerSlot> workersToEvict = this.cluster.getUsedSlotsByTopologyId(topologyEvict.getId());
+        User submitter = this.userMap.get(topologyEvict.getTopologySubmitter());
+
+        LOG.info("Evicting Topology {} with workers: {} from user {}", topologyEvict.getName(), workersToEvict, topologyEvict.getTopologySubmitter());
+        this.nodes.freeSlots(workersToEvict);
+        submitter.moveTopoFromRunningToPending(topologyEvict, this.cluster);
+    }
+
+    private User findUserWithHighestAverageResourceUtilAboveGuarantee() {
+        double most = 0.0;
+        User mostOverUser = null;
+        for (User user : this.userMap.values()) {
+            double over = user.getResourcePoolAverageUtilization() - 1.0;
+            if ((over > most) && (!user.getTopologiesRunning().isEmpty())) {
```

Comment:
So could we directly check the value returned from getResourcePoolAverageUtilization for sanity?
We could still do this in addition to checking whether the user has running topologies if it would be safer, or if it would not really matter, we can remove the user topo check.